### ISSUE TITLE
Crash on develop

### ISSUE THREAD

**Issue by ralight** [eclipse/mosquitto#505]
==16413== Invalid read of size 8
==16413==    at 0x41E29E: tmp_remove_subs (subs.c:553)
==16413==    by 0x41E29E: sub__clean_session (subs.c:606)
==16413==    by 0x40BF3E: context__cleanup (context.c:139)
==16413==    by 0x40C303: context__free_disused (context.c:237)
==16413==    by 0x4118BB: mosquitto_main_loop (loop.c:194)
==16413==    by 0x405199: main (mosquitto.c:363)
==16413==  Address 0xbc97070 is 0 bytes inside a block of size 40 free'd
==16413==    at 0x4C2EDEB: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16413==    by 0x40C8B2: db__message_remove.part.2 (database.c:263)
==16413==    by 0x40D808: db__message_remove (database.c:876)
==16413==    by 0x40D808: db__message_write (database.c:877)
==16413==    by 0x41195E: mosquitto_main_loop (loop.c:247)
==16413==    by 0x405199: main (mosquitto.c:363)
==16413==  Block was alloc'd at
==16413==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==16413==    by 0x4127C5: mosquitto__malloc (memory_mosq.c:69)
==16413==    by 0x40DBDC: db__message_insert (database.c:438)
==16413==    by 0x41BED3: subs__process (subs.c:136)
==16413==    by 0x41C185: sub__search (subs.c:403)
==16413==    by 0x41C08E: sub__search (subs.c:394)
==16413==    by 0x41DFCD: sub__messages_queue (subs.c:528)
==16413==    by 0x40D06A: db__messages_easy_queue (database.c:589)
==16413==    by 0x410ECF: log__vprintf (logging.c:244)
==16413==    by 0x410FF3: log__printf (logging.c:262)
==16413==    by 0x4107A7: handle(short, bool __restrict, short, char, long double, int, bool,...) (handle_subscribe.c:148)
==16413==    by 0x414D28: packet__read (packet_mosq.c:522)

**Comment by toast-uz** [comment]
I've reproduced a crash on develop branch and found out it was an memory issue merged the commit  "Better subtree searching".

Related information is in the conversation of #495 

output of git bisect:
```
883af8af5379092097c6552a7a4a8c52409d2566 is the first bad commit
commit 883af8af5379092097c6552a7a4a8c52409d2566
Author: Roger A. Light <roger@atchoo.org>
Date:   Tue Jul 19 15:05:53 2016 +0100

    Better subtree searching.

:040000 040000 307f444ba3fd19dc7379136fbebce7cf55b55d0d 2bad4ab79806944cb07496f688ad7d58d1e5f9e6 M	src
:040000 040000 6e53242b73223391e0c4c93546c63dc779e68a3d 67c8975f39f7a3d17f87c41a34a94638cba407e4 M	test
```

output of valgrind:
```
$ valgrind --leak-check=full --show-leak-kinds=all src/mosquitto -c ../mosquitto.conf 
==12526== Memcheck, a memory error detector
==12526== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==12526== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==12526== Command: src/mosquitto -c ../mosquitto.conf
==12526== 
==12526== Invalid read of size 8
==12526==    at 0x41C136: tmp_remove_subs (subs.c:553)
==12526==    by 0x41C136: sub__clean_session (subs.c:606)
==12526==    by 0x40B10E: context__cleanup (context.c:137)
==12526==    by 0x40B566: context__free_disused (context.c:231)
==12526==    by 0x41017D: mosquitto_main_loop (loop.c:131)
==12526==    by 0x404888: main (mosquitto.c:360)
==12526==  Address 0x7a225c0 is 0 bytes inside a block of size 104 free'd
==12526==    at 0x4C29CDD: free (vg_replace_malloc.c:530)
==12526==    by 0x4198A6: sub__remove_recurse (subs.c:370)
==12526==    by 0x4197AA: sub__remove_recurse (subs.c:366)
==12526==    by 0x41BB92: sub__remove (subs.c:493)
==12526==    by 0x40F831: handle__unsubscribe (handle_unsubscribe.c:79)
==12526==    by 0x412E40: packet__read (packet_mosq.c:521)
==12526==    by 0x410681: loop_handle_reads_writes (loop.c:486)
==12526==    by 0x410681: mosquitto_main_loop (loop.c:319)
==12526==    by 0x404888: main (mosquitto.c:360)
==12526==  Block was alloc'd at
==12526==    at 0x4C28BE3: malloc (vg_replace_malloc.c:299)
==12526==    by 0x410BA5: mosquitto__malloc (memory_mosq.c:69)
==12526==    by 0x41A2F2: sub__add_hier_entry (subs.c:415)
==12526==    by 0x41B358: sub__add_recurse (subs.c:317)
==12526==    by 0x41B761: sub__add (subs.c:471)
==12526==    by 0x40F5A1: handle(short, bool __restrict, short, char, long double, int, bool,...) (handle_subscribe.c:142)
==12526==    by 0x412E40: packet__read (packet_mosq.c:521)
==12526==    by 0x410681: loop_handle_reads_writes (loop.c:486)
==12526==    by 0x410681: mosquitto_main_loop (loop.c:319)
==12526==    by 0x404888: main (mosquitto.c:360)
(snip)
==12526== LEAK SUMMARY:
==12526==    definitely lost: 2,392 bytes in 23 blocks
==12526==    indirectly lost: 3,640 bytes in 35 blocks
==12526==      possibly lost: 0 bytes in 0 blocks
==12526==    still reachable: 123,502 bytes in 3,503 blocks
==12526==         suppressed: 0 bytes in 0 blocks
==12526== 
==12526== For counts of detected and suppressed errors, rerun with: -v
==12526== ERROR SUMMARY: 71 errors from 6 contexts (suppressed: 0 from 0)
Segmentation fault
```

**Comment by ralight** [comment]
Have you got a reliable easy way of reproducing it you could share?

**Comment by ralight** [comment]
A method is fine, it doesn't have to be code.

**Comment by toast-uz** [comment]
Seemed reproduced when suddenly disconnect by lower subscriber without unsubscribe under hierarchical subscribers. 

You can perfectly reproduce "definitely lost" after running this code for a few seconds and stoping mosquitto manually:
```
package main

import (
	MQTT "github.com/eclipse/paho.mqtt.golang"
)

const optTargetMqtt string = "tcp://localhost:1883"

func main() {
	// subscribe "#" by connection A
	opts := MQTT.NewClientOptions().AddBroker(optTargetMqtt)
	opts.SetClientID("A")
	connA := MQTT.NewClient(opts)
	if token := connA.Connect(); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}
	if token := connA.Subscribe("#", byte(0), nil); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}

	// loop of subscribing "b" and suddenly disconnecting by connection B
	opts.SetClientID("B")
	for {
		connB := MQTT.NewClient(opts)
		if token := connB.Connect(); token.Wait() && token.Error() != nil {
			panic(token.Error())
		}
		if token := connB.Subscribe("b", byte(0), nil); token.Wait() && token.Error() != nil {
			panic(token.Error())
		}
		connB.Disconnect(250)
	}
}
```

You can sometimes reproduce "Invalid read of size 8" during running this code (random test) for a few tens of seconds, and rarely have a crash when you stop this code manually after producing "Invalid read of size 8":
```
package main

import (
	"fmt"
	MQTT "github.com/eclipse/paho.mqtt.golang"
	"math/rand"
	"strings"
	"time"
)

const optTargetMqtt string = "tcp://localhost:1883"
const maxConns int = 100

type status_t int

const (
	DISCONNECT status_t = iota
	CONNECT
	SUBSCRIBE
)

type context struct {
	subtopic string
	status   status_t
	opts     *MQTT.ClientOptions
	conns    MQTT.Client
}

func Loop(i int) {
	// setup context
	var ctx context
	topicletter := []string{"a", "b", "#", ""}
	l := len(topicletter)
	ctx.subtopic = topicletter[i%(l-1)] + topicletter[i/(l-1)%l] + topicletter[i/((l-1)*l)%l] + topicletter[i/((l-1)*l*l)%l]
	if n := strings.Index(ctx.subtopic, "#"); n >= 0 {
		ctx.subtopic = ctx.subtopic[0:n]
	}

	ctx.status = DISCONNECT
	ctx.opts = MQTT.NewClientOptions()
	ctx.opts.AddBroker(optTargetMqtt)
	ctx.opts.SetClientID(fmt.Sprintf("%07d", i))
	r := rand.New(rand.NewSource(time.Now().UnixNano() * int64(i)))

	for {
		p := r.Intn(50)
		switch {
		case ctx.status == DISCONNECT && p < 30:
			// CONNECT
			ctx.conns = MQTT.NewClient(ctx.opts)
			if token := ctx.conns.Connect(); token.Wait() && token.Error() != nil {
				panic(token.Error())
			}
			ctx.status = CONNECT
			fmt.Printf("C")
		case ctx.status == CONNECT && p < 10:
			// DISCONNECT
			ctx.conns.Disconnect(250)
			ctx.status = DISCONNECT
			fmt.Printf("D")
		case ctx.status == CONNECT && 10 <= p && p < 50:
			// SUBSCRIBE
			if token := ctx.conns.Subscribe(ctx.subtopic, byte(0), nil); token.Wait() && token.Error() != nil {
				panic(token.Error())
			}
			ctx.status = SUBSCRIBE
			fmt.Printf("S")
		case ctx.status == SUBSCRIBE && p < 5:
			// DISCONNECT suddenly
			ctx.conns.Disconnect(250)
			ctx.status = DISCONNECT
			fmt.Printf("d")
		case ctx.status == SUBSCRIBE && 5 <= p && p < 10:
			// UNSUBSCRIBE
			if token := ctx.conns.Unsubscribe(ctx.subtopic); token.Wait() && token.Error() != nil {
				panic(token.Error())
			}
			ctx.status = CONNECT
			fmt.Printf("U")
		}
		time.Sleep(time.Duration(r.Intn(1000)) * time.Millisecond)
	}
}

func main() {
	for i := 1; i < maxConns; i++ {
		go Loop(i)
	}
	Loop(0)
}
```

**Comment by toast-uz** [comment]
When I add:
```
	UHPA_FREE_TOPIC(sub);
	mosquitto__free(sub);
```
after ASH_DELETE(hh, parent, sub); in tmp_remove_subs, my 1st test passed.
But my 2nd test still reported memory errors.
In tmp_remove_subs, there seemed invalid but non-null sub->parent. The sub->parent seemed already free in sub__remove_recurse.

**Comment by toast-uz** [comment]
I could find out a patch passed both tests, however I'm not sure it was correct because seemed a sort of symptomatic treatment. There must be some conflicts between subhier's parents and children. My patch cleaned the ghost parents when they removed. Supposed to be prevent the conflicts before they occurred. 
Anyway, I can make a pull-request of my patch in your need.

**Comment by toast-uz** [comment]
I am struggling with this issue.

Investigating struct mosquitto__subhier, I guess  the name of parameters tend to be gotten wrong impression. Subhier->parent is not its parent but like a big brother of the head of hash (Every subhier != subhier->parent->children !!!). Subhier->subs is not a subhier but a leaf.

Some of codes seem to treat parent as a big brother, while others as a real parent. These due to subhier collapsed and memory leaks.

Right ?

First of all, I suggested changing these names (ie. from parent to head, from subs to leafs) and parameter parent should be treated as real parent for the step of fixing this issue.

FYI, a sample dump of subhier: 

```
lv: children generation level
r:  head of hash
f: following item of hash
p: parent of item
c: children of item

lv:0 h:0x22ef400   -> f:0x22ef400   (p:(nil) c:(nil))
lv:0 h:0x22ef400   -> f:0x22ef6d0 $SYS (p:0x22ef400 c:0x22ef860 $SYS)
lv:1 h:0x22ef860 $SYS -> f:0x22ef860 $SYS (p:(nil) c:0x22efb30 broker)
lv:2 h:0x22efb30 broker -> f:0x22efb30 broker (p:(nil) c:0x22efe00 version)
lv:3 h:0x22efe00 version -> f:0x22efe00 version (p:(nil) c:(nil))
lv:3 h:0x22efe00 version -> f:0x22f01d0 timestamp (p:0x22efe00 c:(nil))
lv:3 h:0x22efe00 version -> f:0x22f03f0 uptime (p:0x22efe00 c:(nil))
lv:3 h:0x22efe00 version -> f:0x22f0530 clients (p:0x22efe00 c:0x22f05a0 total)
lv:4 h:0x22f05a0 total -> f:0x22f05a0 total (p:(nil) c:(nil))
lv:4 h:0x22f05a0 total -> f:0x22f0910 inactive (p:0x22f05a0 c:(nil))
lv:4 h:0x22f05a0 total -> f:0x22f0a50 disconnected (p:0x22f05a0 c:(nil))
lv:4 h:0x22f05a0 total -> f:0x22f0b50 active (p:0x22f05a0 c:(nil))
lv:4 h:0x22f05a0 total -> f:0x22f0cb0 connected (p:0x22f05a0 c:(nil))
```

**Comment by toast-uz** [comment]
I guess,  rearrangement of subhier children generations positions $SYS and any root topics to level 0. As a result #501 can be solved.

**Comment by toast-uz** [comment]
First of all, this issue solved by #528 .
I will face #501 in the future.

**Comment by PierreF** [comment]
I think I've found how to reproduce this crash. It seems to be mosquitto__subhier.parent field being sometime used as "sibling"/head of a linked-list.


### Code to reproduce issue
```
import paho.mqtt.client

def on_connect(client, userdata, flags, rc):
    if rc == 0:
        client.subscribe('topic/a', qos=0)
        userdata['do_subscribe'] = True


def on_subscribe(client, userdata, mid, granted_qos):
    if userdata['do_subscribe']:
        client.subscribe('topic/b', qos=0)
        userdata['do_subscribe'] = False
    else:
        client.unsubscribe('topic/a')


def on_unsubscribe(client, userdata, mid):
    client.disconnect()


client = paho.mqtt.client.Client(userdata={})
client.connect("127.0.0.1", 1888)
client.on_connect = on_connect
client.on_subscribe = on_subscribe
client.on_unsubscribe = on_unsubscribe
client.loop_forever()
```

With server started with:
```
$ ./src/mosquitto -p 1888
[...]
Segmentation fault
```
### What client code does

The client does the following action in this order (e.g. wait for completion of each action):
* connect
* subscribe to topic/a
* subscribe to topic/b
* unsubscribe from topic/a
* disconnect

### Why this happen

In short, the mosquitto__subhier of "topic/b" will have parent pointing to mosquitto__subhier of "topic/a". But the later will be free'd by the unsubscribe. So mosquitto__subhier of "topic/b" have a pointer to unallocated memory.

In details, the subscribe on topic/a will create a mosquitto__subhier for "topic/a". During this creation, [when sub__add_hier_entry is called](https://github.com/eclipse/mosquitto/blob/d3239920d789a16cb9c54c598f4fc7d5ef03f2e9/src/subs.c#L317), subhier is the mosquitto__subhier of "topic" and subhier->children is null (no children yet).

Everything work, but parent of "topic/a" is null (when we would probably expect to point "topic"). State at this is:
(mosquitto__subhier for topic)->children = (mosquitto__subhier for topic/a)
(mosquitto__subhier for topic/a)->parent = null

Then, when subscribe on topic/b the [call to sub__add_hier_entry](https://github.com/eclipse/mosquitto/blob/d3239920d789a16cb9c54c598f4fc7d5ef03f2e9/src/subs.c#L317) will have a non null subhier->children. The value is a pointer to mosquitto__subhier for "topic/a".
In [sub__add_hier_entry](https://github.com/eclipse/mosquitto/blob/d3239920d789a16cb9c54c598f4fc7d5ef03f2e9/src/subs.c#L420) this value will be set in the parent field. Which means after this subscription the stat will be:

(mosquitto__subhier for topic)->children = (mosquitto__subhier for topic/a)
(mosquitto__subhier for topic/a)->parent = null
(mosquitto__subhier for topic/b)->parent = (mosquitto__subhier for topic/a)

So we ends up with "topic/b" thinking it's parent is "topic/a".

The unsubscribe of "topic/a" will unallocate the mosquitto__subhier for "topic/a".

Finally when client disconnect, tmp_remove_subs will be called on the mosquitto__subhier of "topic/b" which try to [access the parent field](https://github.com/eclipse/mosquitto/blob/d3239920d789a16cb9c54c598f4fc7d5ef03f2e9/src/subs.c#L552) of it.

**Comment by PierreF** [comment]
PR #531 fix this issue, since it fix the parent value of mosquitto__subhier. For both "topic/a" and "topic/b", the parent will be "topic".

**Comment by toast-uz** [comment]
Thank you for the evaluation.
In addition, I fixed some illegal methods of deleting subhire tree. For example, original subhier_clean collapsed subhire tree by calling HASH_DELETE in order in the loop of HASH_ITER. This did not work, because the target of the first HASH_DELETE was the head of hash and changed unexpectedly the loop parameter '*subhier' to NULL. Therefore, I fixed to keep the head in the loop and deleted it after the loop.
